Materialize singleton aggregate

rdowavic opened the pull request:
Hey Martin,

I was wondering if you could help me understand why it is that when you apply an AstNodeMapper update to a clause, it doesn't propagate to the program. How do I get it to propagate? Otherwise, what's the sensible approach here?

rdowavic committed:
aggregates with any amount of literals in their body are now materialised, including sum x : {B(x)}

rdowavic committed:
aggregates with any amount of literals in their body are now materialised, including sum x : {B(x)}

rdowavic committed:
aggregates with any amount of literals in their body are now materialised, including sum x : {B(x)}

rdowavic committed:
aggregates with any amount of literals in their body are now materialised, including sum x : {B(x)}

rdowavic committed:
trying a minimal example, commented out the first implementation because it causes a segfault :(

rdowavic committed:
trying a minimal example, commented out the first implementation because it causes a segfault :(

rdowavic committed:
trying a minimal example, commented out the first implementation because it causes a segfault :(

rdowavic committed:
implemented the singleton transformer mostly, except I need help to understand AstNodeMapper

rdowavic committed:
implemented the singleton transformer mostly, except I need help to understand AstNodeMapper

rdowavic committed:
implemented the singleton transformer mostly, except I need help to understand AstNodeMapper

rdowavic committed:
it was an AstAtom after all, silly change

rdowavic committed:
it was an AstAtom after all, silly change

rdowavic committed:
it was an AstAtom after all, silly change

rdowavic committed:
first attempt at making the transformer work

rdowavic committed:
first attempt at making the transformer work

rdowavic committed:
first attempt at making the transformer work

rdowavic commented on src/AstTransforms.cpp:
```diff
@@ -222,234 +222,118 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
+	 	std::cout << program << std::endl;	
+		replaceAggregate update(aggregate, variable);
+		foundClause->addToBody(std::unique_ptr<AstLiteral>(aggHead->clone()));
+		foundClause->apply(update);
```
This is where mainly I don't understand why this doesn't work. And I also obviously have to change the setup to this to store multiple  clause/aggregate pairs. How do you recommend doing this? Otherwise, I'm just going to do it with a set of pairs of Aggregate/Clause raw pointers.

mmcgr replied:
You're applying the nodemapper to foundClause, which is actually a clone of the found clause.

rdowavic replied:
Ahhhh Thank you Martin!! Silly!!!

mmcgr replied:
It took me a while to find that problem, but since it did apply the mapping, yet nothing changed, I was suspicious.

rdowavic replied:
Thank you!

mmcgr commented on src/AstTransforms.cpp:
```diff
@@ -221,6 +222,120 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
+            }
+            found = true;
+            foundAggregate.reset(agg.clone());
+            foundClause.reset(clause.clone());
```
```suggestion
            foundClause = const_cast<AstClause*>(&clause););
```

mmcgr replied:
And foundAggregate, of course

mmcgr commented on src/AstTransforms.cpp:
```diff
@@ -221,6 +222,120 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
+    const TypeEnvironment& env = translationUnit.getAnalysis<TypeEnvironmentAnalysis>()->getTypeEnvironment();
+    // collect references to clause / aggregate pairs
+    // let's just make it for one literal for now. Following Martin's approach is a better idea.
+    std::unique_ptr<AstClause> foundClause = nullptr;
+    std::unique_ptr<AstAggregator> foundAggregate = nullptr;
```
```suggestion
    AstClause* foundClause = nullptr;
    AstAggregator* foundAggregate = nullptr;
```

rdowavic commented:
Hey thanks for this!
How do you make sure you don’t add a duplicate variable to a rule? Should I collect up all the variables in a set and then create some synthesised one and then keep incrementing a counter until it’s not in the set anymore, best approach?

Rachel

> On 7 Apr 2020, at 1:47 pm, mmcgr <notifications@github.com> wrote:
> 
> 
> @mmcgr commented on this pull request.
> 
> In src/AstTransforms.cpp <https://github.com/mmcgr/souffle/pull/2#discussion_r404515356>:
> 
> > -					});
> -				});
> -			});
> -
> -		return changed;
> +					}
> +					node->apply(*this);
> +					assert(node != nullptr);
> +					return node;
> +				}
> +		};
> +		std::cout << "before applying update..." << std::endl;
> +	 	std::cout << program << std::endl;	
> +		replaceAggregate update(aggregate, variable);
> +		foundClause->addToBody(std::unique_ptr<AstLiteral>(aggHead->clone()));
> +		foundClause->apply(update);
> You're applying the nodemapper to foundClause, which is actually a clone of the found clause.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/mmcgr/souffle/pull/2#discussion_r404515356>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AE36O2F4A3VBORROUE6MPE3RLKO6HANCNFSM4MCGKSWA>.
>

mmcgr commented:
> Hey thanks for this! How do you make sure you don’t add a duplicate variable to a rule? Should I collect up all the variables in a set and then create some synthesised one and then keep incrementing a counter until it’s not in the set anymore, best approach? Rachel

The set + increment works.  Not sure if it's needed, though. since we already run the UniqueAggregationVariablesTransformer. If that doesn't already cover your needs then it's a good reference.

rdowavic committed:
rewrote AstSingletonAggregateTransformer

rdowavic committed:
rewrote AstSingletonAggregateTransformer

rdowavic committed:
rewrote AstSingletonAggregateTransformer

rdowavic committed:
sorry formatted with clang-format

rdowavic committed:
sorry formatted with clang-format

rdowavic committed:
sorry formatted with clang-format

rdowavic commented:
Hey Martin!
I rewrote the transformer using a set of pairs of aggregates and clauses and it looks like it’s working on the set of examples I’ve tried it on!
Let me know if you see anything wrong, otherwise I think I can finally make a pull request to main souffle.

Rachel

> On 7 Apr 2020, at 4:59 pm, mmcgr <notifications@github.com> wrote:
> 
> 
> Hey thanks for this! How do you make sure you don’t add a duplicate variable to a rule? Should I collect up all the variables in a set and then create some synthesised one and then keep incrementing a counter until it’s not in the set anymore, best approach? Rachel
> 
> The set + increment works. Not sure if it's needed, though. since we already run the UniqueAggregationVariablesTransformer. If that doesn't already cover your needs then it's a good reference.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/mmcgr/souffle/pull/2#issuecomment-610212041>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AE36O2ARGNUF7T6OB7E54YDRLLFM3ANCNFSM4MCGKSWA>.
>

mmcgr commented on src/AstTransforms.cpp:
```diff
@@ -221,6 +222,117 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
+    AstProgram& program = *translationUnit.getProgram();
+    std::set<std::pair<AstAggregator*, AstClause*>> pairs;
+    // collect references to clause / aggregate pairs
+    // let's just make it for one literal for now. Following Martin's approach is a better idea.
```
Don't think this is needed now, especially without knowing what the Martin Approach is.

mmcgr commented on src/AstTransforms.cpp:
```diff
@@ -221,6 +222,117 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
     return changed;
 }
 
+std::string MaterializeSingletonAggregationTransformer::findUniqueVariableName(const AstClause& clause) {
```
The private static helpers could be limited to this cpp file. Just wrap the methods in
```
namespace {
local function
}
```
so that they can only be seen in this translation unit

rdowavic replied:
oh ok. so i am wrapping it in an anonymous namespace?

rdowavic replied:
I changed it to what I think you mean, is it indented right though?

mmcgr replied:
Wrapping it in an anonymous namespace, yes, but that only works if it's a local function - not if it's a class method. Either approach is arguable, especially since the 'local' version is shared with the other transformers in AstTransforms.cpp.

Sooo, stick with your original class version since you prefer that?

rdowavic replied:
Okay that makes sense. Honestly I thought it was a bit inappropriate to put it as a class method cuz its definitely not specific functionality for the class so maybe now i will change it to local so i feel like i have contributed to this process

mmcgr commented on src/AstTransforms.cpp:
```diff
@@ -240,17 +352,17 @@ bool MaterializeAggregationQueriesTransformer::materializeAggregationQueries(
             // -- create a new clause --
 
-            auto relName = "__agg_rel_" + toString(counter++);
+            auto relName = "__agg_body_rel_" + toString(counter++);
```
Haven't you just added a function for this?

rdowavic replied:
Yes, but the function is specific for __agg_rel relations. I wanted to be able to distinguish between what relations were encapsulating an entire aggregate (__agg_rel_) and which ones were just there to materialize the aggregate body (__agg_body_rel_). I could maybe make the function generic for name

mmcgr reviewed (commented):
Looking a lot better, but you still need to merge in upstream. Better yet, do `git rebase -i upstream/master` or whatever your equivalent is.

rdowavic committed:
removed martin comment

rdowavic committed:
removed martin comment

rdowavic committed:
removed martin comment

rdowavic commented:
> Looking a lot better, but you still need to merge in upstream. Better yet, do `git rebase -i upstream/master` or whatever your equivalent is.

I am not sure I have upstream set. should upstream be the main souffle repo?

rdowavic committed:
added namespaces, im not sure if this is right now

rdowavic committed:
added namespaces, im not sure if this is right now

rdowavic committed:
added namespaces, im not sure if this is right now

mmcgr commented:
Several aggregation test fail now, but that's because your transformer was placed before the SemanticChecker which produces the errors. Maybe move it after?

Fixing that leave some more errors, which are weird and I haven't worked out.

rdowavic commented:
Okay I will try that. How do you see the tests again? Do I run it locally or I have to do it through the github pull request?

mmcgr commented:
> > Looking a lot better, but you still need to merge in upstream. Better yet, do `git rebase -i upstream/master` or whatever your equivalent is.
> 
> I am not sure I have upstream set. should upstream the main souffle repo?

It's not so important. Just rebasing to make the commit history look prettier. `rebase -i` lets you select each commit - to edit, to squash into the previous, to reword, or just accept.

rdowavic commented:
> > > Looking a lot better, but you still need to merge in upstream. Better yet, do `git rebase -i upstream/master` or whatever your equivalent is.
> > 
> > 
> > I am not sure I have upstream set. should upstream the main souffle repo?
> 
> It's not so important. Just rebasing to make the commit history look prettier. `rebase -i` lets you select each commit - to edit, to squash into the previous, to reword, or just accept.

who should upstream be? :( I need to set this right? Is upstream my repo or main souffle or your repo?

mmcgr commented:
> who should upstream be? :( I need to set this right? Is upstream my repo or main souffle or your repo?

My workflow has origin as my repo and upstream as souffle-lang. If I do `git fetch --all` then all the links to remote stuff is updated. I can then do `git rebase upstream/master` to rebase onto that - similar to using `git merge`. If I do `git rebase -i ...` then I can edit each commit as I see fit.

This is all to make the history look neater. It's not essential.

mmcgr commented on src/AstTransforms.cpp:
```diff
@@ -221,6 +222,117 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
+        std::string variableName = findUniqueVariableName(*clause);
+        auto* variable = new AstVariable(variableName);
+        // __agg_rel_0(z) :- ...
+        aggHead->addArgument(std::unique_ptr<AstArgument>(variable));
```
This is moving `variable` into the AstAtom - so `variable` is no longer defined. You then use it again to make `equalityLiteral`.

rdowavic replied:
Oh... I thought if I wrap a raw pointer in the unique ptr then if the old raw pointer is still hanging around I can use it for other stuff. This is not possible? I need to clone before I let it go into the aggHead?

mmcgr replied:
I think what I said is wrong. `variable` is still fine, since aggHead still exists. There is a use after delete issue occurring with it, according to the memory sanitiser, but I'm not so sure where, now.

I ran the tests locally, with the memory sanitiser enabled (which is similar to valgrind, if you prefer).
I used
```
make distclean;./bootstrap && ./configure --enable-debug --enable-sanitise-memory && TESTSUITEFLAGS=-j16 SOUFFLE_CONFS=-j1 SOUFFLE_CATEGORY=Semantic make check -j16
```
You would want to remove the category/confs for more general testing.

rdowavic replied:
OHHH I see what you are saying. Once the pointer became wrapped up, maybe later the unique pointer falls out of scope and the variable gets deallocated already. This means maybe I should clone way before

rdowavic replied:
Okay I will try this too!

rdowavic replied:
Okay now I also understand why it can't be that too. ok

rdowavic replied:
> > who should upstream be? :( I need to set this right? Is upstream my repo or main souffle or your repo?
> 
> My workflow has origin as my repo and upstream as souffle-lang. If I do `git fetch --all` then all the links to remote stuff is updated. I can then do `git rebase upstream/master` to rebase onto that - similar to using `git merge`. If I do `git rebase -i ...` then I can edit each commit as I see fit.
> 
> This is all to make the history look neater. It's not essential.

Okay, I don't really get how to connect my fork to the original souffle-lang. How do I actually make it be aware that it exists at all? Do I need like another branch, and that one is souffle-lang? I kind of don't really get it, because upstream is a branch, and so if i set upstream to souffle-lang, doesn't it not make sense bcause souffle-lang has many branches?

mmcgr replied:
Try these git commands to illustrate:
```
git remote
```
This lists all your remote repositories  - if you've only cloned your own repo you will have one called 'origin'.

```
git remote add mmcgr https://github.com/mmcgr/souffle
```
That adds another repo called mmcgr.
```
git fetch mmcgr
```
Now we've got all the branch info from mmcgr
```
git branch -r
```
Lists all the branches in remote repositories

And an alias that you may still have installed from ages ago:
```
git lg
```
to show the git commit tree.

You can only push to a remote repository if you have access, which will generally just be your own repo.

So, now you can add souffle-lang as upstream with
```
git remote add upstream https://github.com/souffle-lang/souffle
git fetch --all
```

rdowavic replied:
Thank you for this!

mmcgr committed:
Revert "added namespaces, im not sure if this is right now"

This reverts commit a0beecf10454a024d67e96fab50418e8b45499d4.

mmcgr committed:
Revert "added namespaces, im not sure if this is right now"

This reverts commit a0beecf10454a024d67e96fab50418e8b45499d4.

mmcgr committed:
Revert "added namespaces, im not sure if this is right now"

This reverts commit a0beecf10454a024d67e96fab50418e8b45499d4.

mmcgr committed:
Remove redundant unique_ptr call

mmcgr committed:
Remove redundant unique_ptr call

mmcgr committed:
Remove redundant unique_ptr call

mmcgr committed:
Reorder transformers so semantic check comes before materialisation

mmcgr committed:
Reorder transformers so semantic check comes before materialisation

mmcgr committed:
Reorder transformers so semantic check comes before materialisation

mmcgr committed:
Refactor to avoid memory issues

mmcgr committed:
Refactor to avoid memory issues

mmcgr committed:
Refactor to avoid memory issues

mmcgr committed:
Stop using pointer after deletion

mmcgr committed:
Stop using pointer after deletion

mmcgr committed:
Stop using pointer after deletion

mmcgr committed:
Merge remote-tracking branch 'upstream/master' into materialize_singleton_aggregate

mmcgr commented:
I added some changes; redoing some raw pointers as unique pointers - mainly to highlight any memory errors I hadn't found yet. I did find a couple of issues, that I've fixed up.

rdowavic commented:
> > who should upstream be? :( I need to set this right? Is upstream my repo or main souffle or your repo?
> 
> My workflow has origin as my repo and upstream as souffle-lang. If I do `git fetch --all` then all the links to remote stuff is updated. I can then do `git rebase upstream/master` to rebase onto that - similar to using `git merge`. If I do `git rebase -i ...` then I can edit each commit as I see fit.
> 
> This is all to make the history look neater. It's not essential.

Okay thank you!!

rdowavic commented:
> I added some changes; redoing some raw pointers as unique pointers - mainly to highlight any memory errors I hadn't found yet. I did find a couple of issues, that I've fixed up.

Okay, thank you! I just pulled this. (Btw, I kind of don't get how it knew to pull your changes, doesn't this just still look like my own fork to my command line? Like, how would it even know about this pull request, since I did it on github?? Because I thought that the .git/config file has everything and it doesn't list your repo, so how does it know about it? Once I have your changes as well, do you think next steps are for me to check it as well and see if there's anything still wrong and then close this pull request and open one on souffle lang?

rdowavic commented:
I see that some of the builds fail. I'm assuming it's possible to reproduce this on my end at least for the Mac OS X ones. I just have to like edit the flags I pass to the compilation stage right? Is this the idea?

rdowavic commented:
> > > who should upstream be? :( I need to set this right? Is upstream my repo or main souffle or your repo?
> > 
> > 
> > My workflow has origin as my repo and upstream as souffle-lang. If I do `git fetch --all` then all the links to remote stuff is updated. I can then do `git rebase upstream/master` to rebase onto that - similar to using `git merge`. If I do `git rebase -i ...` then I can edit each commit as I see fit.
> > This is all to make the history look neater. It's not essential.
> 
> Okay thank you!!

Okay so basically souffle-lang should be listed as a remote. I think I get it now.

mmcgr commented:
> Okay, thank you! I just pulled this. (Btw, I kind of don't get how it knew to pull your changes, doesn't this just still look like my own fork to my command line? Like, how would it even know about this pull request, since I did it on github?? Because I thought that the .git/config file has everything and it doesn't list your repo, so how does it know about it?

If you make a PR to another github repo then by default you give the maintainer access to that branch - so I can push to your repo for that one branch, which I did.

> Once I have your changes as well, do you think next steps are for me to check it as well and see if there's anything still wrong and then close this pull request and open one on souffle lang?

Yes, but as you found there are new problems. :(

A shortened version of semantic/agg_checks3 looks like something like this with your transformer disabled:
```prolog
.decl R1(a:symbol)  
.decl R3(a:symbol,b:number)  
.decl __agg_body_rel_0()  

R1("a").
R1("b").

R3(A1,(1*count : __agg_body_rel_0())) :- 
   R1(A1),
   count : R1(A1) > 1.

__agg_body_rel_0().
.output R3
```
With your transformer:
```prolog
.decl R1(a:symbol)  
.decl R3(a:symbol,b:number)  
.decl __agg_body_rel_0()  
.decl __agg_rel_0(z:number)  
.decl __agg_rel_1(z0:number)  
.decl +disconnected0()  

R1("a").
R1("b").

__agg_body_rel_0().

__agg_rel_0(count : R1(_)).

__agg_rel_1(count : __agg_body_rel_0()).

+disconnected0() :- 
   z > 1,
   __agg_rel_0(z).

R3(A1,(1*z0)) :- 
   +disconnected0(),
   R1(A1),
   __agg_rel_1(z0).
.output R3
```

`count : R1(A1) > 1`  should not be moved.

mmcgr committed:
Check if aggregate vars are anchored outside the aggregate

mmcgr committed:
Check if aggregate vars are anchored outside the aggregate

mmcgr commented:
I added a 'fix' that checks that if a grounded variable in an aggregate is found, the clause containing the aggregate does not have that value. e.g., `C(x) :- A(x), count : B(x)` I'm pretty sure this is an over-approximation.

rdowavic committed:
made changes

rdowavic committed:
made changes

rdowavic commented:
Thanks for pointing this out. I realised my original formulation of when an aggregate was single valued was not covering all the cases. Like you said, whenever a variable that's inside the aggregate also occurs outside, we can't be sure it's single-valued anymore.

Your fix makes sense, but I am worried if the variable occurs twice in the aggregate body but not at all in the outside, it will still be judged as not single-valued.

So I am gunna rewrite this single-valued function. I remember there was a way that you computed the difference of two sets that was a bit more succinct than my formulation, I will have to try to find that again.

rdowavic committed:
rewrote single valued function to be correct instead of wrong

rdowavic commented:
Hey Martin, I'm having trouble running the tests locally.
It says that there's no target to run distclean, is this something I have to set up myself?

mmcgr commented:
`make distclean` only works if you've run `./bootstrap` and `./configure`

`make distclean` followed by `make distclean` fails because the first one removes `Makefile` which tells make how to run distclean.

You can do `git clean -ffxd`, which tries to delete everything not known to git, but be sure that's what you want.

Alternatively, just assume you already cleaned everything if `distclean` fails and move onto `./bootstrap ...`

rdowavic commented:
thank you!

rdowavic commented:
I'm guessing for one of those flags --enable-debug or --enable-sanitise-memory, it requires me to have c++17? I'll go and try to figure out how I can upgrade my version of clang, I guess...

mmcgr commented:
Souffle itself requires C++17 now. Has for a few months. What problem are you having?

rdowavic commented:
Oh okay, I'm not sure, but if I do configure without these flags, it seems to not be bothered by the fact that my gcc ((clang)) has no c++17 support. I'll keep looking around

rdowavic commented:
I tried to see if I could compile a program with -std=c++17 and my clang doesn't seem to mind at all, so I'm confused why in the configure process, it would try all the options:

checking whether g++ supports C++17 features with -std=c++17... no
checking whether g++ supports C++17 features with +std=c++17... no
checking whether g++ supports C++17 features with -h std=c++17... no
checking whether g++ supports C++17 features with -std=c++1z... no
checking whether g++ supports C++17 features with +std=c++1z... no
checking whether g++ supports C++17 features with -h std=c++1z... no
configure: error: *** A compiler with support for C++17 language features is required.

And say that none of them work :(

mmcgr commented:
> I tried to see if I could compile a program with -std=c++17 and my clang doesn't seem to mind at all, so I'm confused why in the configure process, it would try all the options:
> 
> checking whether g++ supports C++17 features with -std=c++17... no
> checking whether g++ supports C++17 features with +std=c++17... no
> checking whether g++ supports C++17 features with -h std=c++17... no
> checking whether g++ supports C++17 features with -std=c++1z... no
> checking whether g++ supports C++17 features with +std=c++1z... no
> checking whether g++ supports C++17 features with -h std=c++1z... no
> configure: error: *** A compiler with support for C++17 language features is required.
> 
> And say that none of them work :(

You can specify your compiler with `CXX=clang++ ./configure`, or whatever your compiler normally is. If this is on your Mac, though, I would expect g++ to be a link to the same compiler anyway but it's worth a try.

rdowavic commented:
Yes, the g++ links to clang :(

mmcgr commented:
You could try updating xcode, if there is an update. I am not a mac person but I think that updates clang.

What is the output for each of these commands?
```
clang++ -v
```
```
g++ -v
```
In case you have this installed via brew:
```
g++-9 -v
```
and configuring souffle using 
```
CXX=clang++ ./configure
```

Commit:
defer groundness check for inlined relations

Commit:
ungrounded inline tests

Commit:
fix ground-chk bailing immediately in semantic-chk

Commit:
Merge remote-tracking branch 'upstream/master' into inline-ground-check

rdowavic commented:
> You could try updating xcode, if there is an update. I am not a mac person but I think that updates clang.
> 
> What is the output for each of these commands?
> 
> ```
> clang++ -v
> ```
> 
> ```
> g++ -v
> ```
> 
> In case you have this installed via brew:
> 
> ```
> g++-9 -v
> ```
> 
> and configuring souffle using
> 
> ```
> CXX=clang++ ./configure
> ```

Yep, I gave this a go:

```[💖] clang++ -v
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin19.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

[💖] g++ -v
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.0 (clang-1100.0.33.17)
Target: x86_64-apple-darwin19.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

[💖]  g++-9 -v
Using built-in specs.
COLLECT_GCC=g++-9
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/9.3.0/libexec/gcc/x86_64-apple-darwin19/9.3.0/lto-wrapper
Target: x86_64-apple-darwin19
Configured with: ../configure --build=x86_64-apple-darwin19 --prefix=/usr/local/Cellar/gcc/9.3.0 --libdir=/usr/local/Cellar/gcc/9.3.0/lib/gcc/9 --disable-nls --enable-checking=release --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-9 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-system-zlib --with-pkgversion='Homebrew GCC 9.3.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --disable-multilib --with-native-system-header-dir=/usr/include --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
Thread model: posix
gcc version 9.3.0 (Homebrew GCC 9.3.0) 

```

Yep, configuring works fine with this CXX=clang++ or without, I can compile souffle, everything is good, it's just as soon as I include --enable-memory-sanitise that it's game over. But I realised this is probably not a big issue, I can just switch over to plang4 when I want to look at the memory stuff.

Maybe I will try to do that now to see what is happening.

mmcgr commented:
Ah, I see now that I missed a message - 

> I'm guessing for one of those flags --enable-debug or --enable-sanitise-memory, it requires me to have c++17? I'll go and try to figure out how I can upgrade my version of clang, I guess...

enable-debug will just work. All that does is use `-g3 -O0` instead of `-O3`. For the sanitiser you could try `CXX=g++-9 ./configure`, use valgrind, or try plang4/8.

Do you need either? Your PR seems to work now.

rdowavic commented:
I noticed that 4 of the tests fail when I run it though (out of 30), but they also seem to fail on my old version of master, should I be concerned? it says there is a free/delete pair with mismatching type :(

Should I give this PR to main souffle now? But I guess I should try to do the whole rebasing thing first.

Commit:
skip semantic checks on clauses w/o relations

mmcgr commented:
> I noticed that 4 of the tests fail when I run it though (out of 30), but they also seem to fail on my old version of master, should I be concerned? it says there is a free/delete pair with mismatching type :(

I think I know the error, and it is fine, and it will also go away when you merge/rebase in master.
 
> Should I give this PR to main souffle now? But I guess I should try to do the whole rebasing thing first.

 I think it's ready for upstream.

Commit:
Merge branch 'master' into inline-ground-check

Commit:
edited incorrect comment

ohamel-softwaresecure committed:
Merge branch 'master' into inline-ground-check

rdowavic committed:
Merge branch 'materialize_singleton_aggregate' of https://github.com/rdowavic/souffle into materialize_singleton_aggregate

rdowavic commented:
Hey Martin, I tried rebasing, but I'm always really afraid of accidentally discarding work or fucking things up somehow. Every single patch that it tried to apply caused a merge conflict and I just skipped through all of them, because I wasn't sure how to deal with it (Sometimes it made me choose between two old versions, which I genuinely don't understand the purpose of, both of them are outdated and I don't want either of them). Now, I've lost your refactored version of the transform function. I am honestly not sure what to do now, besides copying the function body directly from github. Is this all I can do?

rdowavic committed:
fixed rebase

rdowavic commented:
Ok, my final rebased version ended up having some things lost, but I just restored them from the latest github commit. Does this seem fine?

mmcgr commented:
I may have made the rebase seem more important than it was.

There still seems to be some issues. I just tried it myself with my existing branch to see if I could help, but ... it worked automatically for me. I have a branch you could compare to at https://github.com/mmcgr/souffle/tree/theRealMSA

mmcgr commented:
The new git history is quite confusing. If it helps, I can push the branch I linked above to here, which should be almost equivalent to what you had before the rebase. The tip is my change to check if vars are anchored outside the aggregate.

Not that if you're rebasing, and worried about breaking things, you can make a new branch as a cache of the existing, then go back and rebase. Actually, since you're having so much fun with git already, you could even try this to get your pre-rebase branch back:
```sh
cd your/repo
git checkout materialize_singleton_aggregate #If you're not already there
git reflog materialize_singleton_aggregate
#Look puzzled
#Find commit from before the rebase
git reset --hard materialize_singleton_aggregate@{1} #or whatever the appropriate number is
#profit
```

ohamel-softwaresecure committed:
Merge branch 'master' into inline-ground-check

Commit:
Cache total signature during code generation

Commit:
Fix lint warning

Commit:
removed useless comment

Commit:
Merge branch 'inline-ground-check' of github.com:ohamel-softwaresecure/souffle into inline-ground-check

mmcgr committed:
Merge branch 'master' into inline-ground-check

rdowavic committed:
removed duplicate definitions, very silly

rdowavic commented:
Okay that's a good idea, I'll do that next time.
Yeah, I looked into doing that, but I couldn't find the commit and it was confusing :( Next time.
I removed some duplicate definitions that were left (woops, really silly), I think now my thing matches your thing exactly. Once Travis builds and says it's all good, I'll open a pull request on main souffle

mmcgr committed:
Merge branch 'master' into materialize_singleton_aggregate

mmcgr committed:
Merge pull request #1410 from ohamel-softwaresecure/inline-ground-check

Defer groundness check for inlined relations.

mmcgr committed:
Merge branch 'master' into interpreter_total_signature_fix

mmcgr commented:
The diff I have from my out of date backup is:
```diff
diff --git a/src/AstTransforms.cpp b/src/AstTransforms.cpp
index ad02d08d2..9ad673f63 100644
--- a/src/AstTransforms.cpp
+++ b/src/AstTransforms.cpp
@@ -223,7 +223,7 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
 }
 
 std::string MaterializeSingletonAggregationTransformer::findUniqueVariableName(const AstClause& clause) {
-    int counter = 0;
+    static int counter = 0;
     std::set<std::string> variableNames;
     visitDepthFirst(clause, [&](const AstVariable& variable) { variableNames.insert(variable.getName()); });
     std::string candidateVariableName = "z";  // completely arbitrary
@@ -235,7 +235,7 @@ std::string MaterializeSingletonAggregationTransformer::findUniqueVariableName(c
 
 std::string MaterializeSingletonAggregationTransformer::findUniqueAggregateRelationName(
         const AstProgram& program) {
-    int counter = 0;
+    static int counter = 0;
     auto candidate = "__agg_rel_" + toString(counter++);
     while (getRelation(program, candidate) != nullptr) {
         candidate = "__agg_rel_" + toString(counter++);
@@ -317,32 +317,25 @@ bool MaterializeSingletonAggregationTransformer::transform(AstTranslationUnit& t
     return pairs.size() > 0;
 }
 
-// An aggregate is single-valued if none of the variables appearing in the body
-// are ungrounded when ignoring the outer scope
-// ie the aggregate only ever evaluates to a single value.
+// An aggregate is single-valued if
+// any variable occurring inside the aggregate body never occurs in the outer scope.
 bool MaterializeSingletonAggregationTransformer::isSingleValued(
         const AstAggregator& agg, const AstClause& clause) {
-    // Dummy clause to analyse groundedness
-    auto aggClause = std::make_unique<AstClause>();
-    for (const auto& lit : agg.getBodyLiterals()) {
-        aggClause->addToBody(std::unique_ptr<AstLiteral>(lit->clone()));
-    }
-    for (const auto& argPair : getGroundedTerms(*aggClause)) {
-        if (const auto* variable = dynamic_cast<const AstVariable*>(argPair.first)) {
-            // If we have a variable and it's not grounded then we may not be single-valued
-            if (!argPair.second) {
-                return false;
-            }
-            // Variables may get values from outside the aggregate. e.g., A(x), count : B(x)
-            int count = 0;
-            visitDepthFirst(clause, [&](const AstVariable& clauseVariable) {
-                if (*variable == clauseVariable) {
-                    ++count;
-                }
-            });
-            if (count > 1) {
-                return false;
-            }
+    std::map<std::string, int> occurrences;
+    visitDepthFirst(clause, [&](const AstVariable& v) {
+        if (occurrences.find(v.getName()) == occurrences.end()) {
+            occurrences[v.getName()] = 0;
+        }
+        occurrences[v.getName()] = occurrences[v.getName()] + 1;
+    });
+    std::set<std::string> aggVariables;
+    visitDepthFirst(agg, [&](const AstVariable& v) {
+        aggVariables.insert(v.getName());
+        occurrences[v.getName()] = occurrences[v.getName()] - 1;
+    });
+    for (std::string variableName : aggVariables) {
+        if (occurrences[variableName] != 0) {
+            return false;
         }
     }
     return true;
@@ -503,8 +496,8 @@ bool MaterializeAggregationQueriesTransformer::materializeAggregationQueries(
 }
 
 bool MaterializeAggregationQueriesTransformer::needsMaterializedRelation(const AstAggregator& agg) {
-    // everything with more than 1 body literal => materialize
-    if (agg.getBodyLiterals().size() > 1) {
+    // everything with at least 1 body literal => materialize
+    if (agg.getBodyLiterals().size() >= 1) {
         return true;
     }
```

Maybe that very last part? `>= 1`?

rdowavic committed:
the old wrong materialized function fixed

rdowavic committed:
Merge branch 'materialize_singleton_aggregate' of https://github.com/rdowavic/souffle into materialize_singleton_aggregate

rdowavic commented:
> The diff I have from my out of date backup is:
> 
> ```diff
> diff --git a/src/AstTransforms.cpp b/src/AstTransforms.cpp
> index ad02d08d2..9ad673f63 100644
> --- a/src/AstTransforms.cpp
> +++ b/src/AstTransforms.cpp
> @@ -223,7 +223,7 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
>  }
>  
>  std::string MaterializeSingletonAggregationTransformer::findUniqueVariableName(const AstClause& clause) {
> -    int counter = 0;
> +    static int counter = 0;
>      std::set<std::string> variableNames;
>      visitDepthFirst(clause, [&](const AstVariable& variable) { variableNames.insert(variable.getName()); });
>      std::string candidateVariableName = "z";  // completely arbitrary
> @@ -235,7 +235,7 @@ std::string MaterializeSingletonAggregationTransformer::findUniqueVariableName(c
>  
>  std::string MaterializeSingletonAggregationTransformer::findUniqueAggregateRelationName(
>          const AstProgram& program) {
> -    int counter = 0;
> +    static int counter = 0;
>      auto candidate = "__agg_rel_" + toString(counter++);
>      while (getRelation(program, candidate) != nullptr) {
>          candidate = "__agg_rel_" + toString(counter++);
> @@ -317,32 +317,25 @@ bool MaterializeSingletonAggregationTransformer::transform(AstTranslationUnit& t
>      return pairs.size() > 0;
>  }
>  
> -// An aggregate is single-valued if none of the variables appearing in the body
> -// are ungrounded when ignoring the outer scope
> -// ie the aggregate only ever evaluates to a single value.
> +// An aggregate is single-valued if
> +// any variable occurring inside the aggregate body never occurs in the outer scope.
>  bool MaterializeSingletonAggregationTransformer::isSingleValued(
>          const AstAggregator& agg, const AstClause& clause) {
> -    // Dummy clause to analyse groundedness
> -    auto aggClause = std::make_unique<AstClause>();
> -    for (const auto& lit : agg.getBodyLiterals()) {
> -        aggClause->addToBody(std::unique_ptr<AstLiteral>(lit->clone()));
> -    }
> -    for (const auto& argPair : getGroundedTerms(*aggClause)) {
> -        if (const auto* variable = dynamic_cast<const AstVariable*>(argPair.first)) {
> -            // If we have a variable and it's not grounded then we may not be single-valued
> -            if (!argPair.second) {
> -                return false;
> -            }
> -            // Variables may get values from outside the aggregate. e.g., A(x), count : B(x)
> -            int count = 0;
> -            visitDepthFirst(clause, [&](const AstVariable& clauseVariable) {
> -                if (*variable == clauseVariable) {
> -                    ++count;
> -                }
> -            });
> -            if (count > 1) {
> -                return false;
> -            }
> +    std::map<std::string, int> occurrences;
> +    visitDepthFirst(clause, [&](const AstVariable& v) {
> +        if (occurrences.find(v.getName()) == occurrences.end()) {
> +            occurrences[v.getName()] = 0;
> +        }
> +        occurrences[v.getName()] = occurrences[v.getName()] + 1;
> +    });
> +    std::set<std::string> aggVariables;
> +    visitDepthFirst(agg, [&](const AstVariable& v) {
> +        aggVariables.insert(v.getName());
> +        occurrences[v.getName()] = occurrences[v.getName()] - 1;
> +    });
> +    for (std::string variableName : aggVariables) {
> +        if (occurrences[variableName] != 0) {
> +            return false;
>          }
>      }
>      return true;
> @@ -503,8 +496,8 @@ bool MaterializeAggregationQueriesTransformer::materializeAggregationQueries(
>  }
>  
>  bool MaterializeAggregationQueriesTransformer::needsMaterializedRelation(const AstAggregator& agg) {
> -    // everything with more than 1 body literal => materialize
> -    if (agg.getBodyLiterals().size() > 1) {
> +    // everything with at least 1 body literal => materialize
> +    if (agg.getBodyLiterals().size() >= 1) {
>          return true;
>      }
> ```
> 
> Maybe that very last part? `>= 1`?

:( Ah yep, woops, thank you Martin

rdowavic commented:
Yay, it passed!

rdowavic committed:
fixed comment

mmcgr committed:
Merge pull request #1420 from XiaowenHu96/interpreter_total_signature_fix

Cache signature of RamExistenceCheck during IR generation

mmcgr committed:
Merge remote-tracking branch 'upstream/master' into materialize_singleton_aggregate
